### PR TITLE
Extract UploadFilesMetadataBuilder

### PR DESCRIPTION
**Pull request by jcoyne**
## Why was this change made?

So it's possible to upload a bunch of files that have different filenames than their upload paths.
This will be used by h2, where the files are in ActiveJob and have an opaque filename.

## How was this change tested?
Test suite

### CODE REVIEW

File: lib/sdr_client/deposit/upload_files_metadata_builder.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require 'logger'
+
+module SdrClient
+  module Deposit
+    # Collecting all the metadata about the files for a deposit
+    class UploadFilesMetadataBuilder
```
Comment (ndushay) [inline lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: Tests for this class specifically would be awesome, especially the class method.

Also, would it make sense for the instance methods to be private?

**Review by ndushay — CHANGES_REQUESTED** [review]

File: lib/sdr_client/deposit/upload_files_metadata_builder.rb (hunk 1 above)
Comment (jcoyne) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: This class is fully tested by integration tests. Unit tests are brittle and should be used sparingly.  See what Kent C. Dodds (one of the original signers of the agile manifesto) has to say about this https://kentcdodds.com/blog/write-tests

As to the second point. You can't call it from the class method if it's private.

Comment (ndushay) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: I'm not really after unit tests.  More "functional" tests -- what should happen when you call the class method on UploadFilesMetadataBuilder.build -- that is worthy of a test.  That's the basic "consumers of this class should expect" test.

The instance methods other than "build" can be private, no?

Comment (jcoyne) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: @ndushay this already has 100% path and branch coverage.  Adding any more tests are just going to make it harder to manage this project.  Why is it that the exact functionality did it not need tests before, but now that it's written in a functional style it needs them?

Comment (jcoyne) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: The other instance methods could be private, but it's no reason not to have them public either.

Comment (ndushay) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: i don't think it's ridiculous to expect a class to have a test for its public methods.  I understand the functionality was tested before, but I would have to hunt to figure out where the appropriate tests are for the class.  If I change the new class, I wouldn't expect that the ONLY test I would have to update would be for a different class altogether.  I would think "where are the tests for THIS class?   how is it that our tests are so hard to follow?"

However, I see that the sdr-client:
- has no folder for "integration tests"
- already has a bunch of classes without tests (hopefully most are simpler than THIS one, tho)

So I'll approve.

If this is a straight refactor, then perhaps prefix the PR title with "[refactor]"; if this isn't a straight refactor, then I would expect a test for the new functionality.   A quick eyeballing of the code changes in the PR, and your description of the PR ... I wasn't sure.

Further, if I were coming into this code base cold for a tweak in a variety of classes, i would wonder about "um ... where do I add a test for this?"

**Review by ndushay — APPROVED** [review]

File: lib/sdr_client/deposit/upload_files_metadata_builder.rb (hunk 1 above)
Comment (aaron-collier) [reply lib/sdr_client/deposit/upload_files_metadata_builder.rb:8]: My $.02, this looks like pretty much a straight refactor and code climate is reporting full coverage, which tells me it already had coverage and we aren't accidentally removing coverage. So I'm :+1: on this myself.

**Review by aaron-collier — APPROVED** [review]
